flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/DefaultTimeServiceProvider.java

```diff
@@ -28,15 +31,22 @@
  */
 public class DefaultTimeServiceProvider extends TimeServiceProvider {
 
+	/** The containing task that owns this time service provider. */
+	private final StreamTask<?, ?> task;
```

Comment:
If the `StreamTask` is only used for the async exception handling then I think this can be abstracted behind an `AsynchronousExceptionHandler` interface that the task implements. This would have something like `notifyOfAsyncException()` and `checkAsyncExceptions()`. These are the same methods that `StreamTask` has now but we would not expose all of `StreamTask` to the time service provider but have clear-cut interfaces.